### ISSUE TITLE
quantile estimation algo does not enforce some basic invariants

### ISSUE THREAD

**Issue by marccelani** [tdunning/t-digest#113]
when estimating a quantile q, it should hold that est(q) <= est(q+e). 

The quantile estimation algorithm can be vastly inaccurate when there is an a sudden jump in the curve, and this invariant can be broken.

Take the following digest:
{1, 2, 3..19, 1000000} (all weights = 1)

When estimating p90, the desired rank is 18. Based on the algorithm, q=t=18 when we compute the estimate. i=19. delta will be (1M - 18) / 2. The estimated value will be 19 + (0 - 0.5) * delta which is an absurdly negative number.

Recommended fix: We should clamp between m_(i-1) and m_(i+1). At the tails, we should set min to m_(1) when t = 1, and max to m_(m) when t = m.

**Comment by tdunning** [comment]
Yeah. The paper is far behind.

And there is now new code, too. The issue you mention has been dealt with,
I think.

But happy to have another test or five if you have something in mind.

And can you say what version and which suggest you are using?

**Comment by dariomx** [comment]
Hello Dunning,

I felt into the same thing; do you think it could be possible to have an updated paper with the new pseudo-code? I suspect many of us will fall into this trap ;-?

Thanks.

PS: Why is not it published in a peer-reviewed place? Is it due the lack of formal proof? Just curious.

**Comment by tdunning** [comment]
Yes. I am working on an update.

The reason for not publishing in a peer-reviewed place is that I improve
the algorithms at a slightly faster rate than I finish off the paper. And
there is precious little time to work on this. It isn't like this is going
to make my day job product a whole lot better.

I need to draw some lines and get it out. Lack of formal proof is not a
huge blocker in my mind. It would be better with one, but I think that the
simple realization that it is possible to control size easily and
interestingly and that a practically good incremental approximate algorithm
is available should be enough to publish.

**Comment by phongpanat** [comment]
rain for example cannot be used by this method as it may be caused by physical factors at each locatio toward extreme events.

**Comment by tdunning** [comment]
I have no idea what you are talking about. Can you expand your comment a
bit?

**Comment by tdunning** [comment]
OK. So the paper is now up to date.

Your example of {1:19, 1000000} works the way that I think it should and the way that is documented in the main paper. That is, the cdf at 19 exactly is 0.925. If you go an iota above 19, the cdf is 0.95 and below by a skoshi, it drops to 0.9. Going the other way, the quantile from 0.9 to 0.95-epsilon is 19 and from 0.95 to 1, it is 10^6.

This changes a bit if the centroids have more than unit weight. You can see the effect with the following snippet:
```java
        digest = factory(80).create();
        digest.setScaleFunction(ScaleFunction.K_0);

        for (int j = 0; j < 100; j++) {
            for (int i = 1; i < 20; i++) {
                digest.add(i);
            }
            digest.add(1_000_000);
        }
        assertEquals(18, digest.quantile(0.89999999), 0);
        assertEquals(19, digest.quantile(0.9), 0);
        assertEquals(19, digest.quantile(0.949999999), 0);
        assertEquals(1_000_000, digest.quantile(0.95), 0);
```
What is happening here is that I create a digest with the K_0 scaling function so that all centroids are about the same size. I use delta=80 so that we have enough centroids to have a centroid at each integer value (there will also be intermediate centroids). In one test, I get the following last few centroids
```
39 = {Centroid@1120} "Centroid{centroid=17.411764705882355, count=34}"
40 = {Centroid@1121} "Centroid{centroid=18.0, count=35}"
41 = {Centroid@1122} "Centroid{centroid=18.0, count=36}"
42 = {Centroid@1123} "Centroid{centroid=18.651162790697672, count=43}"
43 = {Centroid@1124} "Centroid{centroid=19.0, count=50}"
44 = {Centroid@1125} "Centroid{centroid=476200.42857142846, count=42}"
45 = {Centroid@1126} "Centroid{centroid=1000000.0, count=40}"
46 = {Centroid@1127} "Centroid{centroid=1000000.0, count=40}"
```
This means that `digest.cdf(18)` is very close to 0.875 as it was in the first case and `digest.cdf(19)` is very close to 0.925, but between these values, we get an interpolated cdf so `digest.cdf(18.5)` is close to 0.9.

I think that this gives the behavior you need. What do you think?

**Comment by tdunning** [comment]
I just wanted to say ...

Your question together with two other issues got me to looking more carefully at what happens with repeated values and I actually found a significant bug as a result. Commit coming shortly. I think I improved the behavior with the data you describe previously, but this latest work should substantially improve it further.

Good questions are sometimes almost as good as good answers and your question here was a good one.

The issue had to do with the sort, actually. With K_1, K_2 or K_3 (and variants), it is important that centroids near the edges be increasing in size (near q=0) and then decreasing in size (after the median, headed toward q=1).  This can cause increased error near the median as well.

What was happening is that the sort was only dealing with the means of the centroids. Due to instability in the sort, when there were lots of repeated values, especially near the extremes a heavy centroid would sometimes sort nearer the end than it should be causing very poor accuracy.

I will be checking your specific issues as well to verify that things are better.

Thanks!
